### PR TITLE
Enforce the use of useIsomorphicLayoutEffect to help with server renders

### PR DESCRIPTION
**Pull request by ryanwilsonperkin**
### WHY are these changes introduced?

Fixes warnings that happen during server side render.

### WHAT is this pull request doing?

We have an existing utility called `useIsomorphicLayoutEffect` that's designed according to React's recommendations to avoid warnings when `useLayoutEffect` is invoked during a server side render. This change enforces the use of the utility through Eslint and fixes the few cases where we weren't using it.

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris/blob/main/README.md#local-development)
🗒 [General tophatting guidelines](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md)
📄 [Changelog guidelines](https://github.com/Shopify/polaris/blob/main/.github/CONTRIBUTING.md#changelog)

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
import React from 'react';
import {Page} from '../src';

export function Playground() {
  return (
    <Page title="Playground">
      {/* Add the code you want to test in here */}
    </Page>
  );
}
```

</details>

### 🎩 checklist

- [ ] Tested on [mobile](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting.md#cross-browser-testing)
- [ ] Tested on [multiple browsers](https://help.shopify.com/en/manual/shopify-admin/supported-browsers)
- [ ] Tested for [accessibility](https://github.com/Shopify/polaris/blob/main/documentation/Accessibility%20testing.md)
- [ ] Updated the component's `README.md` with documentation changes
- [ ] [Tophatted documentation](https://github.com/Shopify/polaris/blob/main/documentation/Tophatting%20documentation.md) changes in the style guide

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -78,6 +78,19 @@ module.exports = {
...
     'jsx-a11y/no-noninteractive-element-to-interactive-role': 'off',
+    'no-restricted-imports': [
+      'error',
+      {
+        paths: [
+          {
+            name: 'react',
+            importNames: ['useLayoutEffect'],
+            message:
+              'Please use useIsomorphicLayoutEffect from the utilities director instead',
```
Comment (BPScott) [inline .eslintrc.js]:
```suggestion
              'Please use useIsomorphicLayoutEffect from the utilities directory instead',
```

**Review by alex-page — APPROVED** [review]
Nice thank you @ryanwilsonperkin

**Comment by ryanwilsonperkin** [comment]
@alex-page looks like some unrelated CI is failing, should I merge or wait for any fixes on `main`?

**Comment by alex-page** [comment]
@ryanwilsonperkin ship it and I will fix up the CI.

**Pull request merged**